helper/src/crypto.rs

```diff
@@ -0,0 +1,96 @@
+    100000000000000,     200000000000000,     300000000000000,     400000000000000,     500000000000000,     600000000000000,     700000000000000,     800000000000000,     900000000000000,
+    1000000000000000,    2000000000000000,    3000000000000000,    4000000000000000,    5000000000000000,    6000000000000000,    7000000000000000,    8000000000000000,    9000000000000000,
+    10000000000000000,   20000000000000000,   30000000000000000,   40000000000000000,   50000000000000000,   60000000000000000,   70000000000000000,   80000000000000000,   90000000000000000,
+    100000000000000000,  200000000000000000,  300000000000000000,  400000000000000000,  500000000000000000,  600000000000000000,  700000000000000000,  800000000000000000,  900000000000000000,
+    1000000000000000000, 2000000000000000000, 3000000000000000000, 4000000000000000000, 5000000000000000000, 6000000000000000000, 7000000000000000000, 8000000000000000000, 9000000000000000000,
+    10000000000000000000
+];
+
+/// Runtime initialized H generator.
+static H_PRECOMP: LazyLock<VartimeEdwardsPrecomputation> =
+    LazyLock::new(|| VartimeEdwardsPrecomputation::new([*H, ED25519_BASEPOINT_POINT]));
+
+/// Runtime initialized zero commitment lookup table
+///
+/// ASSUMPTION: This function assume that the [`ZERO_COMMITMENT_DECOMPOSED_AMOUNT`]
+/// table is sorted.
+pub static ZERO_COMMITMENT_LOOKUP_TABLE: LazyLock<[(u64, EdwardsPoint); 172]> =
+    LazyLock::new(|| {
+        let mut lookup_table: [(u64, EdwardsPoint); 172] = [(0_u64, ED25519_BASEPOINT_POINT); 172];
+
+        for (i, amount) in ZERO_COMMITMENT_DECOMPOSED_AMOUNT.into_iter().enumerate() {
+            lookup_table[i] = (
+                amount,
+                (ED25519_BASEPOINT_POINT + *H * Scalar::from(amount)),
+            );
+        }
+
+        lookup_table
+    });
+
+/// This function compute the zero commitment given a specific amount.
+///
+/// It will first attempt to lookup into the table of known Pre-RCT value.
+/// Compute it otherwise.
+pub fn compute_zero_commitments(amount: u64) -> EdwardsPoint {
+    // OPTIMIZATION: We first make an arithmetic check that the value lies between valid pre-computed amounts. This permit to avoid the lookup cost if possible
+    if amount > 10 && amount % 10 != 0 {
+        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::from(1_u8)]);
+    }
+
+    // Binary search in lookup table (O(log2(n)) instead of O(n))
+    let (mut start, mut end, mut c): (usize, usize, Option<EdwardsPoint>) =
+        (0, (ZERO_COMMITMENT_LOOKUP_TABLE.len() - 1) / 2, None);
+    while start <= end {
+        let mid = (start + end) / 2;
+        let lookup = ZERO_COMMITMENT_LOOKUP_TABLE[mid].0;
+        match amount {
+            a if a == lookup => {
+                c = Some(ZERO_COMMITMENT_LOOKUP_TABLE[mid].1);
```

Comment:
I think you are meant to return here, otherwise this will infinite loop.

As a fun exercise I made a "constant time" algorithm:
```rust
    let log = amount.checked_ilog10().unwrap(); // don't actually unwrap here.
    let div = 10_u64.pow(log);

    let most_significant_digit = amount / div;

    if most_significant_digit * div != amount {
        return H_PRECOMP.vartime_multiscalar_mul([Scalar::from(amount), Scalar::ONE]);
    }

    let index = (most_significant_digit + u64::from(log) * 10 - 1 * (u64::from(log) + 1)) as usize;
    
    ZERO_COMMITMENT_LOOKUP_TABLE[index].1
```

note: this is not actually constant time for cryptographic purposes. it would be good to see how that stacks up.


FWIW: I would rather use rusts build in binary search even if it is a little slower.

Reply:
I must say you nailed it. Here's the benchmark:

![constant_time](https://github.com/user-attachments/assets/dd1abfa5-74ec-4100-968e-07b4207d338a)

```
lookup_search/binary_search_manual
                        time:   [835.15 ns 839.11 ns 843.73 ns]
lookup_search/constant_time
                        time:   [871.62 ns 873.04 ns 874.56 ns]
```

The manual implementation of binary search is faster than constant time yet yours is clever (imo).

Reply:
Forget to push the end pointer fix on the manual binary search one:

![fixed](https://github.com/user-attachments/assets/c8dbe51d-5a13-4650-978b-c60442d048c3)

```
lookup_search/binary_search_std
                        time:   [1.2080 µs 1.2101 µs 1.2124 µs]
lookup_search/binary_search_manual
                        time:   [991.33 ns 993.50 ns 995.41 ns]
lookup_search/hashmap_search_std
                        time:   [1.3606 µs 1.3620 µs 1.3636 µs]
lookup_search/constant_time
                        time:   [864.30 ns 865.61 ns 867.13 ns]
```

[Benchmark code.](https://gist.github.com/SyntheticBird45/e97c045aca71c5a0380dba8cf52bd0c5)

I obviously advocate we use your constant time version with a generous documentation on top.

Reply:
constant-time version will be used
https://matrix.to/#/!zPLCnZSsyeFFxUiqUZ:monero.social/$IBmC-hk2AN7XdQwo3T7StkaFDGEeRd-7XBeZenwBUCk?via=monero.social&via=matrix.org&via=kernal.eu